### PR TITLE
Add missing autobuild-disable flag to publish command

### LINKED ISSUE #2391: Porter publish rebuild the image instead of just pushing it
## Describe the bug

My porter bundle is correctly building. But for the build to pass correctly, I need to pass some `--build-arg` (in particular a `GIT_TOKEN`).

During my buid everything runs fine. Then during the publish, for no reason, the `porter publish` command tries to rebuid my image. Because I can't have `--build-arg` at this step, my build fails. 

I'm thinking of a bug here, because just set some retries on my task make it pass and runs correctly (i.e just push the image).

## To Reproduce

1. Run
```bash
curl -L https://cdn.porter.sh/latest/install-linux.sh | bash
export PATH=~/.porter
porter mixins install --url https://github.com/LvffY/ansible-mixin/releases/download ansible
porter mixins install --feed-url https://mchorfa.github.io/porter-helm3/atom.xml helm3
porter mixins install kubernetes
porter build --build-arg GIT_TOKEN="This is a secret for build time !" --build-arg BUNDLE_USER=porter
```
2. Use this porter.yaml & Dockerfile.tmpl

<details><summary>Dockerfile.tmpl</summary>
<p>

```Dockerfile
FROM ubuntu:20.04

## Porter init steps
## See https://porter.sh/bundle/custom-dockerfile/#porter_init
# PORTER_INIT

## Git token to use to clone
ARG GIT_TOKEN
## Mozilla SOPS version
ARG SOPS_VERSION=3.7.3

## Git script to generate password non interactively
## See https://git-scm.com/docs/gitcredentials#_requesting_credentials
ENV GIT_ASKPASS=$BUNDLE_DIR/git_askpass.sh \
    ## Ensure path contains python executable for the porter user
    PATH=$PATH:/home/${BUNDLE_USER}/.local/bin \
    ## Avoid noisy depreciation warnings in python libraries not under our responsabilities
    PYTHONWARNINGS="ignore" \
    ## Ensure Ansible will find installed collections
    ANSIBLE_COLLECTIONS_PATH="$BUNDLE_DIR/collections"

RUN apt-get update && \
    ## Install all necessary tools
    apt-get install -y --no-install-recommends wget gnupg2 python3 python3-dev python3-pip python3-venv git libpq-dev build-essential apt-transport-https ca-certificates &&  \
    ## Install SOPS
    wget https://github.com/mozilla/sops/releases/download/v$SOPS_VERSION/sops_${SOPS_VERSION}_amd64.deb && \
    dpkg -i sops_${SOPS_VERSION}_amd64.deb && \
    ## Ensure linux packages are up-to-date \
    apt-get dist-upgrade -y && \
    ## Because helm mixins install the executable under /usr/local/bin/helm3, we create a symbolic link to avoid too many configurations in our ansible playbooks.
    ## During the installation process, this link is broken.
    ## When helm mixins is installed this link will work.
    ln -sf /usr/local/bin/helm3 /usr/local/bin/helm && \
    ## Clean up image
    apt-get autoremove &&  \
    apt-get autoclean &&  \
    rm -rf /var/lib/apt/lists/* sops_${SOPS_VERSION}_amd64.deb

# Use the BUNDLE_DIR build argument to copy files into the bundle
COPY . $BUNDLE_DIR

## Clone playbook repositories
RUN echo "${GIT_TOKEN:?GIT_TOKEN is not set, please set it.}"

## Porter mixins install
## See https://porter.sh/bundle/custom-dockerfile/#porter_mixins
# PORTER_MIXINS

```

</p>
</details>

<details><summary>porter.yaml</summary>
<p>

```yaml
name: porter/reproduce-porter-error
version: 0.0.9
description: "Bundle to deploy our demo apps"
registry: placeholder
dockerfile: Dockerfile.tmpl
schemaVersion: 1.0.0-alpha.1

mixins:
  - exec
  - ansible:
      otherPipDependencies:
        - virtualenv
        - psycopg2
        - openshift
  ## No explicit use of these mixins but we use this config to don't have to know how to install kubectl & helm
  - kubernetes
  - helm3:
      clientVersion: v3.10.0

parameters:
  - name: git-token
    description: Git token to use to clone ArgoCD repository
    type: string
    env: GIT_TOKEN
    sensitive: true

install:
  - exec:
      description: "Description of the command"
      command: echo # The command to run, must be on the PATH
      arguments: # arguments to pass to the command
        - "Hello install !"

upgrade:
  - exec:
      description: "Description of the command"
      command: echo # The command to run, must be on the PATH
      arguments: # arguments to pass to the command
        - "Hello upgrade !"


uninstall:
  - exec:
      description: "Description of the command"
      command: echo # The command to run, must be on the PATH
      arguments: # arguments to pass to the command
        - "Hello uninstall !"
```

</p>
</details>

3. Run this porter command `...`

```bash
porter publish --registry="my_acr.azurecr.io" --force ## Yes I'm using an ACR, may be the error comes from here ?
```

## Expected behavior

The `porter publish` shoud just push my image and never rebuild it.

## Porter Command and Output

1. porter build command (went well)

[porter-build.txt](https://github.com/getporter/porter/files/9731258/porter-build.txt)

2. porter publish (went wrong then push my previously built image)

[porter-publish.txt](https://github.com/getporter/porter/files/9731293/porter-publish.txt)

## Version
Copy the output of `porter version` below

`porter v1.0.0 (89ad7ed6)`

## Other hypothesis

My image is kind of big (2GB+). May be the issue could some kind of caches that takes a long time to update, so the publish tries to rebuild an unfound image (in its cache) then finally found it after some retries ?

## Azure pipelines

I don't think this a workstation issue, because I'm runnning my builds in Azure pipeline in some Microsoft hosted agents (i.e kind of very general VMs and that could be considered clean every time).

For information, here are the relevant steps on my pipeline : 

<details><summary>azure-pipelines.yaml</summary>
<p>

```yaml
stages:
  - stage: snapshot
    displayName: Build snapshot porter app
    pool:
      vmImage: ubuntu-latest
    jobs:
      - job: snapshot
        displayName: Build snapshot porter app
        steps:
          - checkout: self

          - task: Docker@2
            displayName: Login to ACR
            inputs:
              command: login
              containerRegistry: my_acr

          - task: Bash@3
            displayName: Install porter
            inputs:
              targetType: inline
              script: |
                set -e
                curl -L https://cdn.porter.sh/latest/install-linux.sh | bash
                echo "##vso[task.setvariable variable=PATH]${PATH}:~/.porter"

          - task: Bash@3
            displayName: Porter build
            inputs:
              targetType: inline
              script: |
                set -e
                porter mixins install --url https://github.com/LvffY/ansible-mixin/releases/download ansible
                porter mixins install --feed-url https://mchorfa.github.io/porter-helm3/atom.xml helm3
                porter mixins install kubernetes
                porter build --build-arg GIT_TOKEN=$GIT_TOKEN --build-arg BUNDLE_USER=porter
            env:
              GIT_TOKEN: "This is a secret for build time !"

          - task: Bash@3
            displayName: Porter publish
            retryCountOnTaskFailure: 10
            inputs:
              targetType: inline
              script: |
                set -e
                porter publish --registry="my_acr.azurecr.io" --force

```

</p>
</details>

### PR DESCRIPTION
**Pull request by alekseybb197**
# What does this change

Added cli flag and configuration key for prevent unexpected autobuild invocation image during publish command because it crush pipelines execute.

Use as "porter publish ... --autobuild-disable" in cli or insert "autobuild-disable: true" into config.yaml.

# What issue does it fix
Follow-up to #2573 which missed adding the flag to publish

# Notes for the reviewer

# Checklist
- [ ] Did you write tests?
- [x] Did you write documentation?
- [ ] Did you change porter.yaml or a storage document record? Update the corresponding schema file.
- [ ] If this is your first pull request, please add your name to the bottom of our [Contributors][contributors] list. Thank you for making Porter better! 🙇‍♀️

# Reviewer Checklist
* Comment with /azp run test-porter-release if a magefile or build script was modified
* Comment with /azp run porter-integration if it's a non-trivial PR

[contributors]: https://getporter.org/src/CONTRIBUTORS.md

### CODE REVIEW

**Comment by carolynvs** [comment]
Closing as this was already implemented in https://github.com/getporter/porter/pull/2573. You can find more open issues that are good for new contributors at https://getporter.org/find-issue

**Comment by alekseybb197** [comment]
Sorry, as I saw this flag is absent into publish code! And it crush our pipelines which build and publish cnab. Because we can not able send build-arg when autobuild appeared. But let it be)

**Comment by carolynvs** [comment]
Oops, thanks for pushing back, I thought I had updated the publish command in that pull request and see now that it was missed. I'll reopen and review.

**Comment by carolynvs** [comment]
@alekseybb197 Thank you for the fix! I have pushed 2 commits to your pull request to simplify how --autobuild-disabled is defined on publish and added a test to validate that the flag is working.

Before I can merge, can you push a commit to your branch that includes the following text

```
Signed-off-by: Aleksey Barabanov <Alexey.Barabanov@kaspersky.com>
```

This indicates you agree to our repository's DCO requirement. https://getporter.org/contribute/guide/#signing-your-commits

**Comment by alekseybb197** [comment]
Thanks @carolynvs , I try ...
I was upset and merged as is for my fork)

**Comment by carolynvs** [comment]
Correct, I removed AutoBuildDisabled from datastore.go because it is not necessary. The additional viper configuration in the cmd/porter package and declaring a config setting in DataStore is only needed for configuration values that are **not** defined as flags on a command. So please do not add it back.

You can validate this locally by running `mage build install` and then editing your ~/.porter/config.yaml file to set `autobuild-disabled: true`. Then try to publish a bundle without building it first and you should see it stop immediately because autobuild is disabled.

**Comment by alekseybb197** [comment]
Carolyn, with all due respect, I can't do this.
---------
ansible@test-vm:~/local$ git clone https://github.com/getporter/porter.git
ansible@test-vm:~/local$ cd porter
ansible@test-vm:~/local/porter$ mage build
ansible@test-vm:~/local/porter$ LC_ALL=C ll bin
total 62088
drwxrwx---  5 ansible ansible     4096 Apr 19 23:33 ./
drwxrwxr-x 16 ansible ansible     4096 Apr 19 23:33 ../
drwxrwx---  9 ansible ansible     4096 Apr 19 23:33 mixins/
-rwxrwxr-x  1 ansible ansible 63555130 Apr 19 23:33 porter*
drwxrwx---  2 ansible ansible     4096 Apr 19 23:33 runtimes/
drwxrwx---  2 ansible ansible     4096 Apr 19 23:33 v1.0.13-14-gc0dcb287/
ansible@test-vm:~/local/porter$
ansible@test-vm:~/local/porter$ mage install
installing Porter from bin to /home/ansible/.porter
ansible@test-vm:~/local/porter$ echo "autobuild-disabled: true" >~/.porter/config.yaml
ansible@test-vm:~/local/porter$ cd ../examples/helm-multiarch/
ansible@test-vm:~/local/examples/helm-multiarch$ porter publish -r far.far.away
Changes have been detected and the previously built bundle is out-of-date, rebuilding the bundle before proceeding...
Building bundle ===>
Copying porter runtime ===> 
Copying mixins ===> 
Copying mixin exec ===> 
Copying mixin helm3 ===> 
Building invocation image
[+] Building 1.9s (5/5)                                                                                                                                                
 => [internal] load build definition from Dockerfile                                                                                                              0.0s
 => => transferring dockerfile: 1.26kB                                                                                                                            0.0s
 => [internal] load .dockerignore                                                                                                                                 0.0s
 => => transferring context: 227B                                                                                                                                 0.0s
 => resolve image config for docker.io/docker/dockerfile-upstream:1.4.0                                                                                           1.0s
 => CACHED docker-image://docker.io/docker/dockerfile-upstream:1.4.0@sha256:178c4e4a93795b9365dbf6cf10da8fcf517fcb4a17f1943a775c0f548e9fc2ff                      0.0s
 => [internal] load metadata for docker.io/library/debian:stretch-slim                                                                                            0.7s
^Ccancel requested
------------
Maybe I'm wrong about something. Maybe I'm missing something

**Comment by carolynvs** [comment]
Let me double check locally. I can also update the test to explicitly use config instead of the flag.

**Comment by carolynvs** [comment]
I found a regression in binding config file values to flags in some circumstances. https://github.com/getporter/porter/issues/2735

Once that is fixed (in-progress), then the autobuild-disabled config file setting should work as expected without the additional configuration (defining it on DataStore).

**Comment by carolynvs** [comment]
One of your commits does have a signature, so I have marked the DCO as passing. When I merge, I will squash the commits into a single signed commit, co-authored by both of us. So everything is fine.

I should have that config file regression fixed soon. Once that's merged, I'll update my publish test to use a config file (instead of a flag), and we can be sure that the original issue that you raised is addressed before I merge this PR.

**Comment by alekseybb197** [comment]
Thanks a lot Carolyn! Have a nice day!

File: CONTRIBUTORS.md [hunk 1]
```diff
@@ -80,3 +80,4 @@ and we will add you. **All** contributors belong here. 💯
 * [James Blair](https://github.com/jmhbnz)
 * [Chengwei Guo](https://github.com/cw-Guo)
 * [Sarah Christoff](https://github.com/hypernovasunnix)
+* [Aleksey Barabanov](https://github.com/alekseybb197)
```
Comment (carolynvs) [inline CONTRIBUTORS.md:83]: 🎉

**Review by carolynvs — APPROVED** [review]
Thank you for your patience on this one! I appreciate the bug report and fix 👍

**Pull request merged**